File: src/stepper_motors.cc
#include "stepper_motors.h"

#include "gpio.h"
#include "model.pb.h"
#include "wiring_pi_wrapper.h"
#include <cstdio>
#include <string>

namespace {

void prepare(int dir_pin, const sisyphus::Step_Movement& movement) {
  if (movement == sisyphus::Step::STOP) {
    return;
  }
  Gpio::write(dir_pin, movement == sisyphus::Step::FORWARDS ? LOW : HIGH);
}

void begin(int step_pin, const sisyphus::Step_Movement& movement) {
  if (movement == sisyphus::Step::STOP) {
    return;
  }
  Gpio::write(step_pin, HIGH);
}

void end(int step_pin, const sisyphus::Step_Movement& movement) {
  if (movement == sisyphus::Step::STOP) {
    return;
  }
  Gpio::write(step_pin, LOW);
}

}

void StepperMotors::step(const sisyphus::Step& step) {
  sisyphus::Step_Movement angular_movement = step.angular_movement();
  sisyphus::Step_Movement linear_movement = step.linear_movement();
  prepare(GPIO_STEPPER_MOTORS_ANGLE_DIR, angular_movement);
  prepare(GPIO_STEPPER_MOTORS_LINEAR_DIR, linear_movement);
  begin(GPIO_STEPPER_MOTORS_ANGLE_STEP, angular_movement);
  begin(GPIO_STEPPER_MOTORS_LINEAR_STEP, linear_movement);
  delayMicroseconds(1000);
  end(GPIO_STEPPER_MOTORS_ANGLE_STEP, angular_movement);
  end(GPIO_STEPPER_MOTORS_LINEAR_STEP, linear_movement);
}

File: node/stepper_motors.js
const rpio = require('rpio');
const Gpio = require('./gpio');
const Model = require('./gen/model_pb');

function prepare(dirPin, movement) {
  if (movement == Model.Step.Movement.STOP) {
    return;
  }
  Gpio.write(dirPin, movement == Model.Step.Movement.FORWARDS ? Gpio.LOW : Gpio.HIGH);
}

function begin(stepPin, movement) {
  if (movement == Model.Step.Movement.STOP) {
    return;
  }
  Gpio.write(stepPin, Gpio.HIGH);
}

function end(stepPin, movement) {
  if (movement == Model.Step.Movement.STOP) {
    return;
  }
  Gpio.write(stepPin, Gpio.LOW);
}

function sleepMs(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

class StepperMotors {
  async step(step) {
    const angularMovement = step.getAngularMovement();
    const linearMovement = step.getLinearMovement();
    prepare(Gpio.GPIO_STEPPER_MOTORS_ANGLE_DIR, angularMovement);
    prepare(Gpio.GPIO_STEPPER_MOTORS_LINEAR_DIR, linearMovement);
    begin(Gpio.GPIO_STEPPER_MOTORS_ANGLE_STEP, angularMovement);
    begin(Gpio.GPIO_STEPPER_MOTORS_LINEAR_STEP, linearMovement);
    rpio.msleep(1);
		end(Gpio.GPIO_STEPPER_MOTORS_ANGLE_STEP, angularMovement);
    end(Gpio.GPIO_STEPPER_MOTORS_LINEAR_STEP, linearMovement);
  }
}

module.exports = StepperMotors;
